src/protocol/boolean/dumb_bitwise_add_constant.rs

```diff
@@ -0,0 +1,192 @@
+/// Starting from the least significant digit add up the digits, carrying when required.
+/// We can skip a few multiplications because one input has known values for each bit.
+///
+/// The tricky part is in computing the "carries".
+/// The logic is slightly different for the least significant bit. For that, we just look
+/// at the value of the least significant bit of the constant. If it is a zero, then there
+/// is no way there is any carry, and the answer is just zero. If it is a one, then the carry
+/// will be a one, only if the least significant bit of a `[a_0]_p` is a secret-sharing of one.
+/// So it's just equal to `[a_0]_p`.
+///
+/// For all subsequent carries, once again, it depends on the value of the constant bit at that place:
+/// - when the constant has a bit value of zero - there is only a carry if BOTH of the previous carry
+/// AND the secret bit are shares of one. We can get this result by just multiplying the two.
+///
+/// When the constant has a bit value of one - there is a carry if EITHER the previous carry
+/// OR the secret bit are shares of one. We can get this result using OR (x OR y = x + y - x*y).
+/// So in either case, we need to multiply the previous carry with the next bit of the secret input.
+///
+/// As such, there are a total of l-2 multiplications (one less than the bit-length of the input).
+/// Sometimes, a multiplication can be skipped, because we know, a prioi that the result must be zero.
+pub async fn bitwise_add_constant<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    a: &[S],
+    b: u128,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: ArithmeticSecretSharing<F>,
+{
+    let mut output = Vec::with_capacity(a.len() + 1);
+
+    let mut last_carry_known_to_be_zero = (b & 1) == 0;
+    let mut last_carry = if last_carry_known_to_be_zero {
+        S::ZERO
+    } else {
+        a[0].clone()
+    };
+    let result_bit = if last_carry_known_to_be_zero {
+        a[0].clone()
+    } else {
+        ctx.share_of_one() - &a[0]
+    };
+    output.push(result_bit);
+
+    for (bit_index, bit) in a.iter().enumerate().skip(1) {
+        let mult_result = if last_carry_known_to_be_zero {
+            // TODO: this makes me sad
+            let _ = ctx
+                .narrow(&BitOpStep::from(bit_index))
+                .multiply(record_id, &S::ZERO, &S::ZERO) // this is stupid
+                .await?;
+
+            S::ZERO
+        } else {
+            ctx.narrow(&BitOpStep::from(bit_index))
+                .multiply(record_id, &last_carry, bit)
+                .await?
+        };
+
+        let next_bit_a_one = (b >> bit_index) & 1 == 1;
+        let next_carry = if next_bit_a_one {
+            last_carry_known_to_be_zero = false;
+            -mult_result + &last_carry + bit
+        } else {
+            mult_result
+        };
+
+        let result_bit = if next_bit_a_one {
```

Comment:
Given the math and restructure involved, it would be good to add a comment about what is happening at this specific step